dataprofiler/data_readers/csv_data.py

```diff
@@ -104,21 +104,114 @@ def _determine_delimiter_of_str(data_as_str):
+                cell_type = 'str'
+                try:
+                    float_cell = float(cell)
+                    if int(float_cell) == float_cell:
+                        cell_type = 'int'
+                    else:
+                        cell_type = 'float'
+                except ValueError:
+                    pass
+
+                if cell_type not in ['float', 'int']:
+                    if len(cell) == 0:
+                        cell_type = 'none'
+                    elif cell.isupper():
+                        cell_type = 'upstr'
+                header_check_list[-1].append(cell_type)
+
+        # Flags differences in types between each row (true/false)
+        potential_header = header_check_list[0]
+        differences = []
+        skip_rows = []
+        for i in range(1, len(header_check_list)):
+            differences.append([])
+            # check to see if row should be skiped due to error
+            if len(header_check_list[i]) > len(potential_header):
+                skip_rows.append(i - 1)
+                differences[i - 1] = [False] * len(header_check_list[i])
+            else:
+                for j in range(len(header_check_list[i])):
+                    diff_flag = False
+                    if header_check_list[i][j] != potential_header[j]:
+                        diff_flag = True
+                    differences[i - 1].append(diff_flag)
+            potential_header = header_check_list[i]
+
+        # Predicts the last row that could be the header, given the criteria
+        prior_len = 0
+        row_classic_header_ends = None
+        for i in range(0, len(differences)):
+            # Determine ratio of none in row, must be BELOW threshold
+            none = float(header_check_list[i].count("none")) / float(len(header_check_list[i]))
+            # Determine percent of differences between prior row, must be BELOW threshold
+            diff = float(differences[i].count(True)) / float(len(differences[i]))
+            # Determine percent of string, uppercase string or none in row, must be ABOVE threshold
+            rstr = float((header_check_list[i].count("str")
+                          + header_check_list[i].count("upstr")
+                          + header_check_list[i].count("none"))) / float(len(header_check_list[i]))
+            # Determines if the elements in the row is increasing or decreasing
+            len_increase = False
+            len_not_none = len(header_check_list[i]) - header_check_list[i].count("none")
+            if len_not_none >= prior_len:
+                prior_len = len_not_none
+                len_increase = True
+            # Returns the last row that could reasonably be the header
+            if (rstr > str_thresh and none <= none_thresh and diff >= diff_thresh \
+                    and len_increase and not row_classic_header_ends) \
+                    or (i in skip_rows):
+                row_classic_header_ends = i
```

Comment:
I'm confused, if i is a row that should be skipped, then why would that make this conditional pass and set the return value to i? Could you provide an example to help me better understand this reasoning?

Reply:
so, either we meet the row with different len, or the row with the same len but different value types of cells, we make that row as potential candidate for the header